modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -278,4 +739,349 @@ private ObjectStatisticsImpl aggregateLocalStatistics(
 
         return tblStats;
     }
+
+    /**
+     * Receive and handle statistics propagation message with partitions statistics.
+     *
+     * @param nodeId Sender node id.
+     * @param msg Statistics propagation message with partitions statistics to handle.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    private void receivePartitionsStatistics(UUID nodeId, StatsPropagationMessage msg) throws IgniteCheckedException {
+        UUID locNode = ctx.discovery().localNode().id();
+        for (StatsObjectData partData : msg.data()) {
+            StatsKey key = new StatsKey(partData.key().schema(), partData.key().obj());
+
+            assert partData.type() == StatsType.PARTITION : "Got non partition level statistics by " + key
+                    + " without request";
+
+            if (log.isTraceEnabled())
+                log.trace(String.format("Received partition statistics %s.%s:%d from node %s", key.schema(),
+                        key.obj(), partData.partId(), nodeId));
+
+            GridH2Table tbl = schemaMgr.dataTable(key.schema(), key.obj());
+            if (tbl == null) {
+                if (log.isInfoEnabled())
+                    log.info(String.format("Ignoring outdated partition statistics %s.%s:%d from node %s",
+                            key.schema(), key.obj(), partData.partId(), nodeId));
+
+                continue;
+            }
+            GridDhtPartitionState partState = tbl.cacheContext().topology().partitionState(locNode, partData.partId());
+            if (partState != OWNING) {
+                if (log.isInfoEnabled())
+                    log.info(String.format("Ignoring non local partition statistics %s.%s:%d from node %s",
+                            key.schema(), key.obj(), partData.partId(), nodeId));
+
+                continue;
```

Comment:
Why a MOVING partition ignores incoming statistics from? It is feasible primary or backup.
If first then we will need to recalculate statistics, in other words make the job twice.
If second - we'll loose statistics?

Reply:
Please take a look at current solution. I'll try to scan only local primary partitions. If some partitions wasn't collected on some node - they won't be in parts array in response and sender will reschedule they to some other node.